File: packages/sdk-python/gridseal/adapters/openai.py
"""OpenAI SDK wrapper that auto-captures AI decisions into a GridSeal proof chain."""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from gridseal.core.chain import append_entry
from gridseal.core.hash import canonicalize, sha256
from gridseal.core.storage import StorageAdapter
from gridseal.core.types import (
    AppendEntryInput,
    ChainState,
    DecisionType,
    Err,
    Ok,
    Result,
    err,
    ok,
)

try:
    from openai import AsyncOpenAI
    from openai.types.chat import ChatCompletion
except ImportError:
    AsyncOpenAI = None  # type: ignore[assignment,misc]
    ChatCompletion = None  # type: ignore[assignment,misc]


@dataclass(frozen=True, slots=True)
class CapturedCompletion:
    """Result of a wrapped OpenAI call, containing both the response and the audit entry ID."""

    completion: ChatCompletion
    entry_id: str


@dataclass(frozen=True, slots=True)
class CaptureError:
    """Error from the capture/audit process (the OpenAI call itself succeeded)."""

    type: str
    message: str


@dataclass(slots=True)
class GridSealOpenAI:
    """Wraps an AsyncOpenAI client to auto-capture chat completions into a proof chain.

    Each call to create_completion captures the input, output, model info, and token
    counts into a new proof chain entry stored via the configured storage adapter.
    """

    client: AsyncOpenAI
    chain_id: str
    storage: StorageAdapter
    _chain: ChainState = field(init=False)
    session_id: str | None = None
    actor_id: str | None = None

    def __post_init__(self) -> None:
        self._chain = ChainState(chain_id=self.chain_id)

    async def create_completion(
        self,
        *,
        messages: list[dict[str, Any]],
        model: str = "gpt-4o",
        decision_type: DecisionType | None = "generation",
        tags: dict[str, str] | None = None,
        annotation: str | None = None,
        parent_entry_id: str | None = None,
        **kwargs: Any,  # noqa: ANN401
    ) -> Result[CapturedCompletion, CaptureError]:
        """Make an OpenAI chat completion call and auto-capture an audit entry.

        All extra keyword arguments are forwarded to the OpenAI API.
        """
        # Compute input hash from the canonical form of the messages
        input_payload = {"messages": messages, "model": model, **kwargs}
        input_hash = sha256(canonicalize(input_payload))

        # Make the actual OpenAI API call
        try:
            completion = await self.client.chat.completions.create(
                model=model,
                messages=messages,
                **kwargs,
            )
        except Exception as exc:
            return err(CaptureError(type="OPENAI_API_ERROR", message=str(exc)))

        # Extract output data
        output_data = _extract_completion_data(completion)
        output_hash = sha256(canonicalize(output_data["content"]))

        # Build the audit entry
        entry_id = _generate_entry_id()
        timestamp = datetime.now(tz=timezone.utc).isoformat()

        entry_input = AppendEntryInput(
            entry_id=entry_id,
            timestamp=timestamp,
            entry_type="ai_decision",
            parent_entry_id=parent_entry_id,
            model_id=completion.model,
            model_provider="openai",
            input_hash=input_hash,
            output_hash=output_hash,
            input_token_count=output_data["input_tokens"],
            output_token_count=output_data["output_tokens"],
            decision_type=decision_type,
            session_id=self.session_id,
            actor_id=self.actor_id,
            tags=tags if tags is not None else {},
            annotation=annotation,
        )

        result = append_entry(self._chain, entry_input)
        if isinstance(result, Err):
            return err(
                CaptureError(
                    type="CHAIN_APPEND_ERROR",
                    message=f"Failed to append entry: {result.error}",
                )
            )

        new_chain, entry = result.value
        self._chain = new_chain

        # Persist to storage
        store_result = await self.storage.put_entry(entry)
        if isinstance(store_result, Err):
            return err(
                CaptureError(
                    type="STORAGE_ERROR",
                    message=f"Failed to store entry: {store_result.error}",
                )
            )

        return ok(CapturedCompletion(completion=completion, entry_id=entry_id))

    @property
    def chain(self) -> ChainState:
        """Access the current chain state."""
        return self._chain


def _extract_completion_data(completion: ChatCompletion) -> dict[str, Any]:
    """Extract relevant data from an OpenAI ChatCompletion response."""
    content: str | None = None
    if completion.choices:
        message = completion.choices[0].message
        content = message.content

    input_tokens: int | None = None
    output_tokens: int | None = None
    if completion.usage is not None:
        input_tokens = completion.usage.prompt_tokens
        output_tokens = completion.usage.completion_tokens

    return {
        "content": content,
        "input_tokens": input_tokens,
        "output_tokens": output_tokens,
    }


def _generate_entry_id() -> str:
    """Generate a UUIDv7 for time-ordered entry IDs.

    Falls back to UUIDv4 if uuid7 is not available (Python <3.14).
    """
    try:
        return str(uuid.uuid7())
    except AttributeError:
        return str(uuid.uuid4())


def wrap_openai(
    client: AsyncOpenAI,
    *,
    chain_id: str,
    storage: StorageAdapter,
    session_id: str | None = None,
    actor_id: str | None = None,
) -> GridSealOpenAI:
    """Wrap an AsyncOpenAI client for auto-capture of AI decisions.

    Returns a GridSealOpenAI instance that records every chat completion
    as a proof chain entry.
    """
    if AsyncOpenAI is None:
        raise ImportError(
            "The openai package is required for the OpenAI adapter. "
            "Install it with: pip install gridseal-sdk[openai]"
        )
    return GridSealOpenAI(
        client=client,
        chain_id=chain_id,
        storage=storage,
        session_id=session_id,
        actor_id=actor_id,
    )

File: packages/sdk-node/src/adapters/openai.ts
/**
 * OpenAI SDK wrapper that auto-captures AI decisions into a GridSeal proof chain.
 *
 * Requires the `openai` package as a peer dependency.
 */

import { randomUUID } from "node:crypto";
import type { OpenAI } from "openai";
import type { ChatCompletion, ChatCompletionCreateParamsNonStreaming } from "openai/resources/chat/completions";
import {
  type AppendEntryInput,
  type ChainState,
  type DecisionType,
  type ProofChainEntry,
  type Result,
  type StorageAdapter,
  type StorageError,
  appendEntry,
  canonicalize,
  createChain,
  ok,
  err,
  sha256,
} from "@gridseal/core";

/** Result of a wrapped OpenAI call, containing both the response and the audit entry ID. */
export type CapturedCompletion = {
  readonly completion: ChatCompletion;
  readonly entryId: string;
};

/** Error from the capture/audit process (the OpenAI call itself may have succeeded). */
export type CaptureError = {
  readonly type: "OPENAI_API_ERROR" | "CHAIN_APPEND_ERROR" | "STORAGE_ERROR";
  readonly message: string;
};

/** Options for creating a GridSealOpenAI wrapper. */
export type GridSealOpenAIOptions = {
  readonly client: OpenAI;
  readonly chainId: string;
  readonly storage: StorageAdapter;
  readonly sessionId?: string | null;
  readonly actorId?: string | null;
};

/** Options for a single completion call. */
export type CreateCompletionOptions = {
  readonly messages: ReadonlyArray<ChatCompletionCreateParamsNonStreaming["messages"][number]>;
  readonly model?: string;
  readonly decisionType?: DecisionType | null;
  readonly tags?: Readonly<Record<string, string>>;
  readonly annotation?: string | null;
  readonly parentEntryId?: string | null;
} & Omit<ChatCompletionCreateParamsNonStreaming, "messages" | "model" | "stream">;

/** Wraps an OpenAI client to auto-capture chat completions into a proof chain. */
export type GridSealOpenAI = {
  /** Make an OpenAI chat completion call and auto-capture an audit entry. */
  readonly createCompletion: (
    options: CreateCompletionOptions
  ) => Promise<Result<CapturedCompletion, CaptureError>>;
  /** Access the current chain state. */
  readonly getChain: () => ChainState;
};

function extractCompletionData(completion: ChatCompletion): {
  content: string | null;
  inputTokens: number | null;
  outputTokens: number | null;
} {
  let content: string | null = null;
  if (completion.choices.length > 0) {
    const firstChoice = completion.choices[0];
    if (firstChoice !== undefined) {
      content = firstChoice.message.content;
    }
  }

  let inputTokens: number | null = null;
  let outputTokens: number | null = null;
  if (completion.usage !== undefined) {
    inputTokens = completion.usage.prompt_tokens;
    outputTokens = completion.usage.completion_tokens;
  }

  return { content, inputTokens, outputTokens };
}

function generateEntryId(): string {
  return randomUUID();
}

/** Create a GridSealOpenAI wrapper around an OpenAI client. */
export function wrapOpenAI(options: GridSealOpenAIOptions): GridSealOpenAI {
  const {
    client,
    chainId,
    storage,
    sessionId = null,
    actorId = null,
  } = options;

  let chain: ChainState = createChain(chainId);

  const createCompletion = async (
    callOptions: CreateCompletionOptions
  ): Promise<Result<CapturedCompletion, CaptureError>> => {
    const {
      messages,
      model = "gpt-4o",
      decisionType = "generation",
      tags,
      annotation = null,
      parentEntryId = null,
      ...extraParams
    } = callOptions;

    const inputPayload: Record<string, unknown> = {
      messages,
      model,
      ...extraParams,
    };
    const inputHash = sha256(canonicalize(inputPayload));

    let completion: ChatCompletion;
    try {
      completion = await client.chat.completions.create({
        model,
        messages: [...messages],
        ...extraParams,
      });
    } catch (error: unknown) {
      const message = error instanceof Error ? error.message : String(error);
      return err({ type: "OPENAI_API_ERROR", message });
    }

    const outputData = extractCompletionData(completion);
    const outputHash = sha256(canonicalize(outputData.content));

    const entryId = generateEntryId();
    const timestamp = new Date().toISOString();

    const entryInput: AppendEntryInput = {
      entryId,
      timestamp,
      entryType: "ai_decision",
      parentEntryId,
      modelId: completion.model,
      modelProvider: "openai",
      inputHash,
      outputHash,
      inputTokenCount: outputData.inputTokens,
      outputTokenCount: outputData.outputTokens,
      decisionType,
      sessionId,
      actorId,
      tags: tags ?? {},
      annotation,
    };

    const appendResult = appendEntry(chain, entryInput);
    if (!appendResult.ok) {
      return err({
        type: "CHAIN_APPEND_ERROR",
        message: `Failed to append entry: ${JSON.stringify(appendResult.error)}`,
      });
    }

    chain = appendResult.value.chain;
    const entry: ProofChainEntry = appendResult.value.entry;

    const storeResult: Result<ProofChainEntry, StorageError> =
      await storage.putEntry(entry);
    if (!storeResult.ok) {
      return err({
        type: "STORAGE_ERROR",
        message: `Failed to store entry: ${JSON.stringify(storeResult.error)}`,
      });
    }

    return ok({ completion, entryId });
  };

  const getChain = (): ChainState => chain;

  return { createCompletion, getChain };
}
